pkg/mcs/resource_manager/client/client.go

```diff
@@ -0,0 +1,774 @@
+	OnResponse(ctx context.Context, resourceGroupName string, req RequestInfo, resp ResponseInfo) error
+}
+
+// ResourceGroupProvider provides some api to interact with resource manager server。
+type ResourceGroupProvider interface {
+	ListResourceGroups(ctx context.Context) ([]*rmpb.ResourceGroup, error)
+	GetResourceGroup(ctx context.Context, resourceGroupName string) (*rmpb.ResourceGroup, error)
+	AddResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceGroup) (string, error)
+	ModifyResourceGroup(ctx context.Context, metaGroup *rmpb.ResourceGroup) (string, error)
+	DeleteResourceGroup(ctx context.Context, resourceGroupName string) (string, error)
+	AcquireTokenBuckets(ctx context.Context, request *rmpb.TokenBucketsRequest) ([]*rmpb.TokenBucketResponse, error)
+}
+
+var _ ResourceGroupKVInterceptor = (*resourceGroupsController)(nil)
+
+type resourceGroupsController struct {
+	clientUniqueID   uint64
+	provider         ResourceGroupProvider
+	groupsController sync.Map
+	config           *Config
+
+	calculators []ResourceCalculator
+
+	// tokenResponseChan receives token bucket response from server.
+	// And it handles all resource group and runs in main loop
+	tokenResponseChan chan []*rmpb.TokenBucketResponse
+
+	// lowTokenNotifyChan receives chan notification when the number of available token is low
+	lowTokenNotifyChan chan struct{}
+
+	run struct {
+		now             time.Time
+		lastRequestTime time.Time
+
+		// requestInProgress is true if we are in the process of sending a request.
+		// It gets set to false when we receives the response in the main loop,
+		// even in error cases.
+		requestInProgress bool
+
+		// requestNeedsRetry is set if the last token bucket request encountered an
+		// error. This triggers a retry attempt on the next tick.
+		//
+		// Note: requestNeedsRetry and requestInProgress are never true at the same time.
+		requestNeedsRetry bool
+
+		// targetPeriod indicate how long it is expected to cost token when acquiring token.
+		// last update.
+		targetPeriod time.Duration
+	}
+}
+
+// NewResourceGroupController returns a new resourceGroupsController which impls ResourceGroupKVInterceptor
+func NewResourceGroupController(clientUniqueID uint64, provider ResourceGroupProvider, requestUnitConfig *RequestUnitConfig) (*resourceGroupsController, error) {
+	var config *Config
+	if requestUnitConfig != nil {
+		config = generateConfig(requestUnitConfig)
+	} else {
+		config = DefaultConfig()
+	}
+	return &resourceGroupsController{
+		clientUniqueID:     clientUniqueID,
+		provider:           provider,
+		config:             config,
+		lowTokenNotifyChan: make(chan struct{}, 1),
+		tokenResponseChan:  make(chan []*rmpb.TokenBucketResponse, 1),
+		calculators:        []ResourceCalculator{newKVCalculator(config), newSQLCalculator(config)},
+	}, nil
+}
+
+// Start starts resourceGroupController service
+func (c *resourceGroupsController) Start(ctx context.Context) error {
```

Comment:
How about add a `Stop` function, I may use it when resource group features disable.

Reply:
I like it